File: starters/flutter/lib/src/core/offline/offline_hook.dart
const offlineHookContractVersion = 'offline-hook/2.0.0';

abstract interface class OfflineHook {
  String get contractVersion;
  bool get enabled;
  Future<String> enqueue({required String id, required String kind});
}

final class DisabledOfflineHook implements OfflineHook {
  const DisabledOfflineHook();

  @override
  String get contractVersion => offlineHookContractVersion;

  @override
  bool get enabled => false;

  @override
  Future<String> enqueue({required String id, required String kind}) async {
    return 'disabled';
  }
}

File: starters/react-native/src/offline/offline-hook.ts
export const OFFLINE_HOOK_CONTRACT_VERSION = 'offline-hook/2.0.0';

export interface OfflineHook {
  readonly contractVersion: typeof OFFLINE_HOOK_CONTRACT_VERSION;
  readonly enabled: boolean;
  enqueue(operation: Readonly<{ id: string; kind: string }>): Promise<'queued' | 'disabled'>;
}

export function createDisabledOfflineHook(): OfflineHook {
  return Object.freeze({
    contractVersion: OFFLINE_HOOK_CONTRACT_VERSION,
    enabled: false,
    enqueue: async (): Promise<'disabled'> => 'disabled',
  });
}
